dotnet/src/Agents/Abstractions/ChatHistoryChannel.cs

```diff
@@ -25,16 +26,47 @@ public class ChatHistoryChannel : AgentChannel
             throw new KernelException($"Invalid channel binding for agent: {agent.Id} ({agent.GetType().FullName})");
         }
 
-        await foreach (ChatMessageContent message in historyHandler.InvokeAsync(this._history, cancellationToken).ConfigureAwait(false))
+        int messageCount = this._history.Count;
+        HashSet<ChatMessageContent> mutatedHistory = [];
+
+        Queue<ChatMessageContent> messageQueue = [];
+        ChatMessageContent? yieldMessage = null;
+        await foreach (ChatMessageContent responseMessage in historyHandler.InvokeAsync(this._history, cancellationToken).ConfigureAwait(false))
+        {
+            for (int messageIndex = messageCount; messageIndex < this._history.Count; messageIndex++)
+            {
+                ChatMessageContent mutatedMessage = this._history[messageIndex];
+                mutatedHistory.Add(mutatedMessage);
+                messageQueue.Enqueue(mutatedMessage);
+            }
+
+            messageCount = this._history.Count;
+
+            if (!mutatedHistory.Contains(responseMessage))
+            {
+                this._history.Add(responseMessage);
+                messageQueue.Enqueue(responseMessage);
+            }
+
+            yieldMessage = messageQueue.Dequeue();
+            yield return (IsMessageVisible(yieldMessage), yieldMessage);
+        }
```

Comment:
I think it would be beneficial for current me and future you to add some comments in this section :). I think there is some underlying understanding of how `_history` is mutated during the async iteration that may not be clear to readers.

Reply:
Agree entirely.